### ISSUE TITLE
Greedy/maximize repair option

### ISSUE THREAD

**Issue by mrmin123** [perryhuynh/kcauto#155]
~~Depending on playstyle and setup it may be advantageous to either repair any ship or only fleet ship. Add in the ability to switch to either or.~~

See discussion below.

**Comment by mrmin123** [comment]
I've been mulling over the logic for this the past few days and I figured I should ask the users' feedback on this. I've had requests for a more 'greedy' repair logic, but as my current playstyle is very compatible with the current fleet-only repair logic, I'm at a bit of a loss as to what is actually being requested.

* The currently implemented logic in kcauto-kai searches and repairs only the damaged ships in the current combat fleet(s) on the first page of the repair list (hereafter referenced as **fleet-only**)
* My initial plan for this 'greedy' logic was to make it like kancolle-auto's [original repair logic](https://github.com/mrmin123/kancolle-auto/blob/master/kancolle_auto.sikuli/combat.sikuli/combat.py#L447): repair the known number of damaged ships in the combat fleet(s), but not necessarily the ships in the combat fleets. If the incorrect ships were repaired, the repair logic would be run again after the next failed sortie attempt (hereafter referenced as **dumb-greedy**). A rough implementation of this already [exists in-code](https://github.com/mrmin123/kcauto-kai/blob/master/kcauto-kai.sikuli/repair.sikuli/repair.py#L203) but is not implemented and exists as a stub.

The current fleet-only implementation falls apart if the player has enough damaged ships in their docks that the fleet ships are pushed off the first page. @waicool20 made a hybrid repair logic that basically combines the dumb-greedy stub implementation and the fleet-only implementation, basically selecting any ship to repair if a fleet ship is not found on the first page, theoretically working away at the damaged ships until the fleet ships are reached. My longer-term solution to the issue was intended to be #154.

That said, there are multiple ways to flesh out the repair logic moving forward, depending on what type of logic is actually desired for the people who need this more greedy repair logic:

1. Implement #154. Does this supplant the need for a greedier repair logic (this ticket)?
2. Use the hybrid logic made by @waicool20. The logic to select fleet ships until they can no longer be chosen on the first page, then falling back to repairing any ship means that eventually the ships in the fleet will be reached. Does this supplant the need for #154?
3. Implement #154 *and* a hybrid method. Select fleet ships beyond the first page, up to a certain page, and then fall back to selecting any random ship to repair?
4. Does 'greedy' actually mean 'maximize repair dock usage'? AKA use fleet-only implementation (with or without #154), but if there are repair docks left after repairing the fleet ships, fill the remaining repair slots with non-fleet ships as well?

Please let me know which logic is the desired one, or write down your own desired logic here or on Discord.

edit: please note that the ability to switch different logics might be configurable, if there is demand for it.

**Comment by waicool20** [comment]
Posting this here for those who don't want to login and aren't in the discord servers: https://strawpoll.com/5rp3hg67

**Comment by twisting2017** [comment]
1 & 4 as configureable options would be the best solution imho. another thing that would be nice would be if ships could be repaired that are damaged but not yet damaged enough to get the 'light damage' tag.

**Comment by mrmin123** [comment]
Judging by the poll, yes, options 1 and 4 are the most desired behavior.

I'm going to modify this ticket to encompass option 4 (greedy), and bump up the priority in option 1 and probably work on that first.

I'll consider repairing scratch damage but that might end up being a separate ticket and not make it in until post-1.0.

**Comment by mrmin123** [comment]
I'll also probably eventually implement something that repairs all damaged ships in the port while not sortieing (a passive repair, so to speak).

**Comment by waicool20** [comment]
Did some research based on @twisting2017 request of repairing non-minor damage ships and I came up with the following code so far, I reference the same functions that I have posted in #168 `similarity()` and `getAverageColor()`, once again pardon the kotlin heh:
```
import java.awt.Color
import java.nio.file.Files
import java.nio.file.Paths
import java.text.DecimalFormat
import javax.imageio.ImageIO

fun main(args: Array<String>) {
    Files.walk(Paths.get("")).filter { it.toString().matches(Regex("Damage.*png")) }.sorted().forEach {
        println("--- $it")
        val image = ImageIO.read(it.toFile())
        val color = image.getAverageColor()
        println("Red: ${color.red}\tGreen: ${color.green}\tBlue: ${color.blue}")
        println(String.format("Hex: #%02X%02X%02X", color.red, color.green, color.blue))

        val pixels = mutableListOf<Double>()
        for (x in 0 until 158) {
            val similarity = Color(image.getRGB(x, 4), true).similarity(Color.WHITE)
            if (pixels.takeLast(4).all { it > 0.9 } && pixels.size > 4) {
                pixels.add(1.0)
            } else {
                pixels.add(similarity)
            }
        }

        val percentHealth = 1 - (pixels.count { it > 0.9 }) / 158.0
        println(DecimalFormat("Estimated Health: #.##%").format(percentHealth))
        println()
    }
}
```
Output: 
```
--- Damage100.png
Red: 96	Green: 250	Blue: 93
Hex: #60FA5D
Estimated Health: 98.73%

--- Damage83.png
Red: 186	Green: 249	Blue: 118
Hex: #BAF976
Estimated Health: 81.65%

--- Damage86.png
Red: 171	Green: 249	Blue: 113
Hex: #ABF971
Estimated Health: 84.81%

--- DamageChuuha.png
Red: 230	Green: 195	Blue: 133
Hex: #E6C385
Estimated Health: 50.63%

--- DamageShouha.png
Red: 217	Green: 222	Blue: 104
Hex: #D9DE68
Estimated Health: 68.99%

--- DamageShouhaOrange.png
Red: 230	Green: 203	Blue: 134
Hex: #E6CB86
Estimated Health: 50%

--- DamageTaiha.png
Red: 228	Green: 179	Blue: 170
Hex: #E4B3AA
Estimated Health: 20.89%
```
Images are based on 158x7 pixel screencapture of each health bar as below (follows the above order):
![damage100](https://user-images.githubusercontent.com/5538666/40364360-bd2e9f28-5e04-11e8-8f4a-9c5bb1e44c9e.png)
![damage86](https://user-images.githubusercontent.com/5538666/40364359-bd0091f0-5e04-11e8-9b52-cbf83aac1eba.png)
![damage83](https://user-images.githubusercontent.com/5538666/40364357-bcd379fe-5e04-11e8-9932-b854fa5854a0.png)
![damagechuuha](https://user-images.githubusercontent.com/5538666/40364361-bd8e434c-5e04-11e8-8dd5-a39b18e5fe9b.png)
![damageshouha](https://user-images.githubusercontent.com/5538666/40364362-bdbb6958-5e04-11e8-931d-faf75bba53b3.png)
![damageshouhaorange](https://user-images.githubusercontent.com/5538666/40364363-bde6619e-5e04-11e8-974e-aa8329f1a6b1.png)
![damagetaiha](https://user-images.githubusercontent.com/5538666/40364365-be3ca95a-5e04-11e8-9c9f-f872dff4a2eb.png)

`getAverageColor()` is not really needed here but I just added it for some insight. For the most part the estimation looks pretty accurate provided that the screen capture is accurately taken. Luckily these health bars don't move around, so it should be easy.

The code works by going through the middle row of the image and calculating its similarity to the color white with a threshold. If it detects more than 4 pixels being white in a row it will just assume the rest of the other pixels are white, to get rid of inaccuracy due to the damage status icon. 

Finally the estimated health is just calculated by dividing by number of pixels in the image.

**Comment by mrmin123** [comment]
Since #154 is implemented I will be punting this to post-1.0.0.

**Comment by mrmin123** [comment]
As of [v8.0.0-rc1](https://github.com/mrmin123/kcauto/releases/tag/v8.0.0-rc1) kcauto now has Passive Repair capabilities.
